TestKit Documentation

renatocaval opened the pull request.

patriknw commented on docs/src/main/paradox/testing.md:
```diff
@@ -0,0 +1,45 @@
+
+### Transitive dependencies
+
+The table below shows `akka-projection-cassandra`'s direct dependencies and the second tab shows all libraries it depends on transitively.
```
akka-projection-testkit

patriknw commented on examples/src/test/java/jdocs/testkit/TestKitDocExample.java:
```diff
@@ -0,0 +1,24 @@
+
+    public TestKitDocExample() {
+        //#projection-testkit
+        ActorTestKit actorTestKit = ActorTestKit.create("testing-projection");
```
better to show this in a Junit test using TestKitJunitResource, see CassandraProjectionTest

renatocaval replied:
I was explicitly trying to avoid bringing in test framework integration to stay independent and show the basics. 

But fine, I don't have strong opinions on that.

patriknw replied:
Sure, we can take that approach, and refer to Akka's testing documentation.

patriknw commented on examples/src/test/scala/docs/cassandra/CassandraProjectionDocExample.scala:
```diff
@@ -40,18 +42,32 @@ import scala.concurrent.Future
         case ShoppingCart.CheckedOut(cartId, time) =>
           logger.info("Shopping cart {} was checked out at {}", cartId, time)
+          repository.save(CartCheckout(cartId, time))
           Future.successful(Done)
```
return the Future of the save

patriknw commented on examples/src/test/scala/docs/testkit/TestKitDocExample.scala:
```diff
@@ -0,0 +1,36 @@
+class TestKitDocExample extends TestSuite with Matchers with BeforeAndAfterAll with ScalaFutures {
+
+  //#projection-testkit
+  val actorTestKit = ActorTestKit("testing-projection")
```
better to show it with ScalaTestWithActorTestKit, since that is probably what they will use

seglo commented on docs/src/main/paradox/testing.md:
```diff
@@ -0,0 +1,67 @@
+# Testing
+
+Akka Projection provides a Test Kit to ease testing. There are two supported styles of test: running with an assert function and driving it with a Akka Streams TestKit `TestSink`.
```
```suggestion
Akka Projection provides a Test Kit to ease testing. There are two supported styles of test: running with an assert function and driving it with an Akka Streams TestKit `TestSink`.
```

seglo commented on docs/src/main/paradox/testing.md:
```diff
@@ -0,0 +1,67 @@
+
+When testing with an assert function the Projection is started and stopped by the test kit. While the projection is running, the assert function will be called until it completes without errors (no exceptions or assertion errors are thrown). 
+
+In the example bellow the Projection will update a Shopping Cart Checkout view. The test will run until it observes that the shopping cart for id `abc-def` was checkout.  
```
```suggestion
In the example below the Projection will update a Shopping Cart Checkout view. The test will run until it observes that the shopping cart for id `abc-def` was checked out.  
```

seglo commented on docs/src/main/paradox/testing.md:
```diff
@@ -0,0 +1,67 @@
+Java
+:  @@snip [TestKitDocExample.java](/examples/src/test/java/jdocs/testkit/TestKitDocExample.java) { #fixme }
+
+Differently then testing with assert function  
```
```suggestion
Differently than testing with assert function  
```

Incomplete sentence.

renatocaval commented:
I reworked the samples for the TestKit. Instead of trying to reuse code from the other sample, I just added new 'fake' implementation for the TestKit. That make it mush easier to follow. 

The code is not runnable and I think it should, but we can fix it in another iteration so we get that part of the docs ready for the first release.

patriknw commented on examples/src/test/java/jdocs/testkit/TestKitDocExample.java:
```diff
@@ -0,0 +1,143 @@
+      cartCheckoutRepository
+              .findById("abc-def")
+              // use a probe to capture the async result
+              .thenAccept(view -> testProbe.ref().tell(view));
```
that can be simplifed by not using a `TestProbe`. Something like
```
CartView cartView = cartCheckoutRepository..findById("abc-def").toCompletableFuture().get(1, TimeUnit.SECONDS);
assertEquals("abc-def", cartView.id);
```

renatocaval replied:
well, I was exactly trying to avoid this nasty api, but agree it aligns better with the ScalaTest `futureValue`

renatocaval replied:
ah, I remember now another issue. 

Calling `.get()` throws exception and because we are inside a Runnable we have to handle it in place (instead of letting it bubble).

patriknw replied:
ok, then we leave it as this now, but we should change `run` to take one of Akka's function types that allows throwing Exception

renatocaval replied:
ok, I can explore it now and send in the PR with the fix for `runWithTestSink`. For after release
